### ISSUE TITLE
Conflict between DDP and flash-attention when key-dim!=value-dim

### ISSUE THREAD

**Issue by artnoage** [Dao-AILab/flash-attention#632]
For some reason although it is supported, I get an error when I try to have these two values different. The error went away with
with torch.backends.cuda.sdp_kernel(enable_math=False), but to be honest, I don't know if I am doing something wrong this way. Any input/ideas are welcome.

**Comment by tridao** [comment]
We don't support key_dim != value_dim.
It's uncommon I think, and we don't want to increase the number of templates too much. Right now we already template on the head dimension, and having different head dimensions between Q/K and V would increase the number of templates (and hence compile time) by a large factor.

**Comment by artnoage** [comment]
Thanks for answering. I know it is uncommon. I am experimenting with different architectures having smaller dimension for K and Q. My initial way to reduce dim of Q,K was to use LoRa and it actually seemed promising. But actual reduction of KQ dimension will also speed up training apart for making the model smaller. Anyway I will think of something. 

I am confused about something though; in one of the repos I use, the function torch.nn.functional.scaled_dot_product_attention appears  which requires "flash". As I said in my initial message this works with different size if you put this math=False option.  I initially thought that this function is an older/stable version of flash attention. Is there actually any relationship between this library and yours?

**Comment by tridao** [comment]
torch SDPA does call our code when things are compatible (dtype, device, etc.). I'm not sure what code is run in this particular case, but we've never supported key_dim != value_dim. It's possible SDPA is falling back to some pure pytorch implementation.

**Comment by artnoage** [comment]
Thanks again!
